### PR TITLE
Configurable traceback modes

### PR DESCRIPTION
**Pull request by ogbanugot**
Run with default trace mode `full`
```
import ivy
import jax.numpy as jnp

ivy.set_backend('jax')
x = jnp.array([])
y = jnp.array([1., 3., 4.])
ret = ivy.functional.frontends.torch.div(x, y)

  File "/Users/ogbanugot/Workspace/ivy/ivy/exceptions.py", line 123, in new_fn
    return fn(*args, **kwargs)
  File "/Users/ogbanugot/Workspace/ivy/ivy/func_wrapper.py", line 415, in new_fn
    return fn(*args, **kwargs)
  File "/Users/ogbanugot/Workspace/ivy/ivy/func_wrapper.py", line 360, in new_fn
    return fn(*args, **kwargs)
  File "/Users/ogbanugot/Workspace/ivy/ivy/func_wrapper.py", line 84, in new_fn
    return fn(*new_args, **new_kwargs)
  File "/Users/ogbanugot/Workspace/ivy/ivy/func_wrapper.py", line 146, in new_fn
    ret = fn(*args, **kwargs)
  File "/Users/ogbanugot/Workspace/ivy/ivy/functional/backends/jax/elementwise.py", line 144, in divide
    ret = jax.numpy.divide(x1, x2)
  File "/Users/ogbanugot/.conda/envs/ivy_dev3/lib/python3.8/site-packages/jax/_src/numpy/ufuncs.py", line 234, in true_divide
    return lax.div(x1, x2)

During the handling of the above exception, another exception occurred:

  File "execpttest.py", line 7, in <module>
    ret = ivy.functional.frontends.torch.div(x, y)
  File "/Users/ogbanugot/Workspace/ivy/ivy/functional/frontends/torch/func_wrapper.py", line 81, in new_fn
    ret = fn(*args, **kwargs)
  File "/Users/ogbanugot/Workspace/ivy/ivy/functional/frontends/torch/func_wrapper.py", line 68, in new_fn
    return fn(*new_args, **new_kwargs)
  File "/Users/ogbanugot/Workspace/ivy/ivy/functional/frontends/torch/pointwise_ops.py", line 284, in div
    return ivy.divide(input, other, out=out)
  File "/Users/ogbanugot/Workspace/ivy/ivy/exceptions.py", line 129, in new_fn
    raise ivy.exceptions.IvyBackendException(fn.__name__, str(e))

IvyBackendException: jax: divide: div got incompatible shapes for broadcasting: (0,), (3,).
```

Run with trace mode `full` but squash `func_wrapper.py`
```
import ivy
import jax.numpy as jnp

ivy.set_backend('jax')
ivy.set_show_func_wrapper_trace_mode(False)
x = jnp.array([])
y = jnp.array([1., 3., 4.])
ret = ivy.functional.frontends.torch.div(x, y)

<func_wrapper.py stack trace is squashed, call `ivy.set_show_func_wrapper_traces(True)` in order to view this>
  File "/Users/ogbanugot/Workspace/ivy/ivy/exceptions.py", line 123, in new_fn
    return fn(*args, **kwargs)
  File "/Users/ogbanugot/Workspace/ivy/ivy/functional/backends/jax/elementwise.py", line 144, in divide
    ret = jax.numpy.divide(x1, x2)
  File "/Users/ogbanugot/.conda/envs/ivy_dev3/lib/python3.8/site-packages/jax/_src/numpy/ufuncs.py", line 234, in true_divide
    return lax.div(x1, x2)

During the handling of the above exception, another exception occurred:

<func_wrapper.py stack trace is squashed, call `ivy.set_show_func_wrapper_traces(True)` in order to view this>
  File "execpttest.py", line 8, in <module>
    ret = ivy.functional.frontends.torch.div(x, y)
  File "/Users/ogbanugot/Workspace/ivy/ivy/functional/frontends/torch/pointwise_ops.py", line 284, in div
    return ivy.divide(input, other, out=out)
  File "/Users/ogbanugot/Workspace/ivy/ivy/exceptions.py", line 129, in new_fn
    raise ivy.exceptions.IvyBackendException(fn.__name__, str(e))

IvyBackendException: jax: divide: div got incompatible shapes for broadcasting: (0,), (3,).
```

Run with trace mode `frontend` show `func_wrapper.py`
```
import ivy
import jax.numpy as jnp

ivy.set_backend('jax')
ivy.set_exception_trace_mode("frontend")
ivy.set_show_func_wrapper_trace_mode(True)
x = jnp.array([])
y = jnp.array([1., 3., 4.])
ret = ivy.functional.frontends.torch.div(x, y)

<stack trace is truncated to frontend specific files, call `ivy.set_exception_trace_mode('full')` to view the full trace>

During the handling of the above exception, another exception occurred:

<stack trace is truncated to frontend specific files, call `ivy.set_exception_trace_mode('full')` to view the full trace>
  File "/Users/ogbanugot/Workspace/ivy/ivy/functional/frontends/torch/func_wrapper.py", line 81, in new_fn
    ret = fn(*args, **kwargs)
  File "/Users/ogbanugot/Workspace/ivy/ivy/functional/frontends/torch/func_wrapper.py", line 68, in new_fn
    return fn(*new_args, **new_kwargs)
  File "/Users/ogbanugot/Workspace/ivy/ivy/functional/frontends/torch/pointwise_ops.py", line 284, in div
    return ivy.divide(input, other, out=out)

IvyBackendException: jax: divide: div got incompatible shapes for broadcasting: (0,), (3,).
```

### CODE REVIEW

File: ivy/exceptions.py
Comment (djl11) [inline ivy/exceptions.py:64]: we shouldn't repeatedly call `ivy.get_exception_trace_mode()`. We should call this once and then store the result in a variable called `trace_mode`.

**Review by djl11 — CHANGES_REQUESTED** [review]
small change, but could you also please implement some unit tests for these various modes? :)

**Comment by ogbanugot** [comment]
Hey @djl11, so there are merge conflicts now because of the PR for `func_wrapper` mode Julia added. I'll fix the conflicts and also make sure all the modes are working fine together. Once that is done I'll request a review again. Thanks!

File: ivy_tests/test_ivy/test_exceptions.py
Comment (djl11) [inline ivy_tests/test_ivy/test_exceptions.py:5]: the tests shouldn't simply check that the modes are set correctly, they should also raise exceptions and verify that the stack traces are what is expected. Basically you should create ground truth stack trace strings and assert that the stack traces produced are what we expect. This will also make it easy for me to verify that things are working as expected :)

**Review by djl11 — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_exceptions.py
Comment (ogbanugot) [reply ivy_tests/test_ivy/test_exceptions.py:5]: Oh okay, I see. I'll do that.

Comment (juliagsy) [reply ivy_tests/test_ivy/test_exceptions.py:5]: Hey! Probably because I'm the assignee so I got the related notifications aha

Anyways, I have an idea / suggestion, will it work if we test that certain keywords of the stack trace is in the traces (and also test fails if keywords that are not supposed to be there is included in the traces)

for example, if the mode is `frontend`, keywords like "backends", "site-packages" (which refer to the actual backend framework library, because asserting "torch" or "jax" for example, may also refer to frontends which can be acceptable in some cases) will cause the test to fails because they belongs to the backend traces and should not appear in the stack trace string

just a basic idea aha, hope this helps and feel free to modify or improve based on this, or implement a better one if you now have one! 

Thanks!

Comment (ogbanugot) [reply ivy_tests/test_ivy/test_exceptions.py:5]: Hey @juliagsy yeah I think this could work! Thanks! Although, we won't have ground truths anymore which I believe would have been a sure way for @djl11 to know that things work exactly as expected. But I do think your suggestion is better than no tests at all!

**Review by djl11 — APPROVED** [review]
I haven't dive tooo deep, as I don't know this area of the code so well. @juliagsy knows it better than I do, so let's wait for her review as well :) If she gives the green light, it's good to merge I think!

File: ivy_tests/test_ivy/test_exceptions.py
Comment (juliagsy) [inline ivy_tests/test_ivy/test_exceptions.py:34]: would it be better to put the first 2 asserts outside of the conditional block, and `assert "/site-packages" in lines` if `backend not in [...]` to simplify the code?

**Review by juliagsy — COMMENTED** [review]
Hey Ogban! Everything looks good! Just some feedback on simplifying some of the code blocks, feel free to merge after that (provided that the tests pass)! Thanks a lot!

**Pull request merged**
